platform/posix/transport/src/mbedtls_pkcs11_posix.c

```diff
@@ -0,0 +1,958 @@
+                        mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                        mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+            returnStatus = MBEDTLS_PKCS11_INTERNAL_ERROR;
+        }
+    }
+
+    if( returnStatus == MBEDTLS_PKCS11_SUCCESS )
+    {
+        /* Set the underlying IO for the TLS connection. */
+        mbedtls_ssl_set_bio( &( pMbedtlsPkcs11Context->context ),
+                             ( void * ) &( pMbedtlsPkcs11Context->socketContext ),
+                             mbedtls_net_send,
+                             mbedtls_net_recv,
+                             mbedtls_net_recv_timeout );
+
+        returnStatus = configureMbedtlsFragmentLength( pMbedtlsPkcs11Context );
+    }
+
+    if( returnStatus != MBEDTLS_PKCS11_SUCCESS )
+    {
+        contextFree( pMbedtlsPkcs11Context );
+    }
+    else
+    {
+        LogInfo( ( "(Network connection %p) TLS handshake successful.", pNetworkContext ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static MbedtlsPkcs11Status_t configureMbedtlsCertificates( MbedtlsPkcs11Context_t * pMbedtlsPkcs11Context,
+                                                           const MbedtlsPkcs11Credentials_t * pMbedtlsPkcs11Credentials )
+
+{
+    MbedtlsPkcs11Status_t returnStatus = MBEDTLS_PKCS11_SUCCESS;
+    int32_t mbedtlsError = 0;
+    CK_RV result = CKR_OK;
+
+    /* Parse the server root CA certificate into the SSL context. */
+    mbedtlsError = mbedtls_x509_crt_parse_file( &( pMbedtlsPkcs11Context->rootCa ),
+                                                pMbedtlsPkcs11Credentials->pRootCaPath );
+
+    if( mbedtlsError != 0 )
+    {
+        LogError( ( "Failed to parse server root CA certificate: mbedTLSError= %s : %s.",
+                    mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                    mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+        returnStatus = MBEDTLS_PKCS11_INVALID_CREDENTIALS;
+    }
+    else
+    {
+        mbedtls_ssl_conf_ca_chain( &( pMbedtlsPkcs11Context->config ),
+                                   &( pMbedtlsPkcs11Context->rootCa ),
+                                   NULL );
+        /* Setup the client private key. */
+        result = initializeClientKeys( pMbedtlsPkcs11Context,
+                                       pMbedtlsPkcs11Credentials->pPrivateKeyLabel );
+
+        if( result != CKR_OK )
+        {
+            LogError( ( "Failed to setup key handling by PKCS #11." ) );
+            returnStatus = MBEDTLS_PKCS11_INVALID_CREDENTIALS;
+        }
+    }
+
+    if( returnStatus == MBEDTLS_PKCS11_SUCCESS )
+    {
+        /* Setup the client certificate. */
+        mbedtlsError = readCertificateIntoContext( pMbedtlsPkcs11Context,
+                                                   pMbedtlsPkcs11Credentials->pClientCertLabel,
+                                                   CKO_CERTIFICATE,
```

Comment:
Why does this parameter need to be sent when the function mentions that it is meant for reading only certificate types from PKCS11?